src/api/c/pinverse.cpp

```diff
@@ -0,0 +1,170 @@
+    Array<char> cond = logicOp<T, af_ge_t>(sVecCast, relTolArr,
+                                           sVecCast.dims());
+    Array<T> zeros = createValueArray<T>(sVecCast.dims(), scalar<T>(0.));
+    sVecRecip = createSelectNode<T>(cond, sVecRecip, zeros, sVecRecip.dims());
+    sVecRecip.eval();
+
+    // Make s vector into s pinverse array
+    Array<T> sPinv = diagCreate<T>(sVecRecip, 0);
+
+    Array<T> uT = transpose(u, true);
+
+    // Crop v and u* for final matmul later based on s+'s size, because
+    // sVec produced by svd() has minimal dim length (no extra zeroes).
+    // Thus s+ produced by diagCreate() will have minimal dims as well,
+    // and v could have an extra dim0 or u* could have an extra dim1
+    if (v.dims()[1] > sPinv.dims()[0]) {
+        vector<af_seq> seqs = {
+            {0., static_cast<double>(v.dims()[0] - 1), 1.},
+            {0., static_cast<double>(sPinv.dims()[0] - 1), 1.}
+        };
+        v = createSubArray<T>(v, seqs);
+    }
+    if (uT.dims()[0] > sPinv.dims()[1]) {
+        vector<af_seq> seqs = {
+            {0., static_cast<double>(sPinv.dims()[1] - 1), 1.},
+            {0., static_cast<double>(uT.dims()[1] - 1), 1.}
+        };
+        uT = createSubArray<T>(uT, seqs);
+    }
+
+    Array<T> out = matmul<T>(matmul<T>(v, sPinv, AF_MAT_NONE, AF_MAT_NONE),
+                             uT, AF_MAT_NONE, AF_MAT_NONE);
+
+    return out;
+}
+
+template<typename T>
+static inline af_array pinverse(const af_array in, const double tol)
+{
+    Array<T> inArray = getArray<T>(in);
+    uint batchSize = inArray.dims()[2];
+
+    if (inArray.ndims() < 3) {
+        return getHandle(pinverseSvd<T>(getArray<T>(in), tol));
+    }
+    else {
+        vector<Array<T> > outputs;
+        for (int i = 0; i < batchSize; ++i) {
+            vector<af_seq> seqs = {
+                {0., static_cast<double>(inArray.dims()[0] - 1), 1.},
+                {0., static_cast<double>(inArray.dims()[1] - 1), 1.},
+                {static_cast<double>(i), static_cast<double>(i), 1.}
+            };
+            Array<T> inSlice = createSubArray<T>(inArray, seqs);
+            outputs.push_back(pinverseSvd<T>(inSlice, tol));
+        }
+        Array<T> mergedOuts = join<T>(2, outputs);
+        return getHandle(mergedOuts);
+    }
+}
+
+af_err af_pinverse(af_array *out, const af_array in, const double tol,
+                   const af_mat_prop options)
+{
+    try {
+        const ArrayInfo& i_info = getInfo(in);
+
+        if (i_info.ndims() > 3) {
+            AF_ERROR("solve can not be used in batch mode", AF_ERR_BATCH);
+        }
+
+        af_dtype type = i_info.getType();
+
+        if (options != AF_MAT_NONE) {
+            AF_ERROR("Using this property is not yet supported in inverse", AF_ERR_NOT_SUPPORTED);
+        }
+
+        ARG_ASSERT(1, i_info.isFloating()); // Only floating and complex types
+        ARG_ASSERT(2, tol >= 0.); // Only floating and complex types
```

Comment:
The comment is incorrect here.

Reply:
Whoops copy paste. Just changed the comment.